Add call stack management (#3651)
ROCm-Developer-Tools/HIP: docs/how-to/hip_runtime_api/call_stack.rst

.. meta::
    :description: This page describes call stack concept in HIP
    :keywords: AMD, ROCm, HIP, call stack

*******************************************************************************
Call stack
*******************************************************************************

The call stack is a data structure for managing function calls, by saving the
state of the current function. Each time a function is called, a new call frame
is added to the top of the stack, containing information such as local
variables, return addresses and function parameters. When the function
execution completes, the frame is removed from the stack and loaded back into
the corresponding registers. This concept allows the program to return to the
calling function and continue execution from where it left off.

The call stack for each thread must track its function calls, local variables,
and return addresses. However, in GPU programming, the memory required to store
the call stack increases due to the parallelism inherent to the GPUs. NVIDIA
and AMD GPUs use different approaches. NVIDIA GPUs have the independent thread
scheduling feature where each thread has its own call stack and effective

> program counter. On AMD GPUs threads are grouped; each warp has its own call

Reviewer: I thought AMD used wavefront instead of warp?

                    << hipGetErrorString(status) \
                    << " at " << __FILE__ << ":" \
                    << __LINE__ << std::endl;    \
        }                                        \
    }

    int main()
    {
        size_t stackSize;
        HIP_CHECK(hipDeviceGetLimit(&stackSize, hipLimitStackSize));
        std::cout << "Default stack size: " << stackSize << " bytes" << std::endl;

        // Set a new stack size
        size_t newStackSize = 1024 * 8; // 8 KiB
        HIP_CHECK(hipDeviceSetLimit(hipLimitStackSize, newStackSize));

        HIP_CHECK(hipDeviceGetLimit(&stackSize, hipLimitStackSize));
        std::cout << "Updated stack size: " << stackSize << " bytes" << std::endl;

        return 0;
    }

Handling recursion and deep function calls
-------------------------------------------------------------------------------

Similar to CPU programming, recursive functions and deeply nested function
calls are supported. However, developers must ensure that these functions do
not exceed the available stack memory, considering the huge amount of memory
needed for the call stack due to the GPUs inherent parallelism. This can be

> achieved by increasing stack size, implementing error handling to catch stack

Reviewer 2: "Error handling to catch stack overflow" is not a solution to exceeding the call stack. At most it can prevent the call stack to corrupt other memory on the GPU, but that also is not under the direct control of the developer. It can inform the user/developer about a stack overflow, but there is no way to recover from a stack overflow (at least not an easy way, and I'm not aware of one on GPUs). A stack overflow is a bug that needs to be fixed.

overflow, optimizing code to reduce stack usage, and utilizing profiling tools
to monitor stack memory. Proper kernel design and memory management strategies

> are essential to maintain efficient and stable application performance.

Reviewer 2: This is not about "efficient and stable [...] performance", but about not crashing due to a stack overflow. Also "memory management strategies" sounds like it's referring to actual controllable allocations on the device. The developer does not have direct control over the call stack

Reviewer 3: Removed this line.

.. meta::
    :description: This page describes call stack concept in HIP
    :keywords: AMD, ROCm, HIP, call stack

*******************************************************************************
Call stack
*******************************************************************************

> The call stack is a concept in computer science, representing the stack data

Reviewer: ```suggestion
The call stack is a data
```

> structure that used for managing function calls, by saving the state of the

Reviewer: ```suggestion
structure for managing function calls, by saving the state of the
```

current function. Each time a function is called, a new call frame is added to
the top of the stack, containing information such as local variables, return
addresses and function parameters. When the function execution completes, the
frame is removed from the stack and loaded back into the corresponding
registers. This concept allows the program to return to the calling function
and continue execution from where it left off.

The call stack for each thread must track its function calls, local variables,
and return addresses. However, in GPU programming, the memory required to store
the call stack increases due to the parallelism inherent to the GPUs. NVIDIA

> and AMD GPUs are using different approaches. NVIDIA GPUs have the independent

Reviewer: ```suggestion
and AMD GPUs use different approaches. NVIDIA GPUs have the independent
```

> thread scheduling feature since Volta GPU, where every thread has its own call

Reviewer: ```suggestion
thread scheduling feature where each thread has its own call
```

stack and effective program counter. On AMD GPUs threads are grouped;

> therefore, every warp/wavefront has its own call stack and program counter, but

Reviewer: ```suggestion
each warp/wavefront has its own call stack and program counter, but
```

Reviewer: Also, in general do we need both warp and wavefront here? Don't we use the term wavefront?

Author: In the other part of the documentation, the term warp is more frequently used than wavefront. I added a link to the relevant section of the Programming Model to help clarify the differences from NVIDIA's terminology.

> not each individual thread.

Reviewer: ```suggestion
not each thread.
```

If a thread or warp exceeds its stack size, a stack overflow occurs, causing
kernel failure. This can be detected using debuggers.

Call stack management with HIP
===============================================================================

> Developers can adjust the call stack size, allowing fine-tuning based on

Reviewer: ```suggestion
You can adjust the call stack size as shown in the following example, allowing fine-tuning based on
```

    #include <hip/hip_runtime.h>
    #include <iostream>

    #define HIP_CHECK(expression)                \
    {                                            \
        const hipError_t status = expression;    \
        if(status != hipSuccess){                \
                std::cerr << "HIP error "        \
                    << status << ": "            \
                    << hipGetErrorString(status) \
                    << " at " << __FILE__ << ":" \
                    << __LINE__ << std::endl;    \
        }                                        \
    }

    int main()
    {
        size_t stackSize;
        HIP_CHECK(hipDeviceGetLimit(&stackSize, hipLimitStackSize));
        std::cout << "Default stack size: " << stackSize << " bytes" << std::endl;

        // Set a new stack size
        size_t newStackSize = 1024 * 8; // 8 KiB
        HIP_CHECK(hipDeviceSetLimit(hipLimitStackSize, newStackSize));

        HIP_CHECK(hipDeviceGetLimit(&stackSize, hipLimitStackSize));
        std::cout << "Updated stack size: " << stackSize << " bytes" << std::endl;

        return 0;
    }

Depending on the GPU model, at full occupancy, it can consume a significant

> amount of memory. For instance, an MI300X with 304 computing units (CU) and up

Reviewer: ```suggestion
amount of memory. For instance, an MI300X with 304 compute units (CU) and up
```

> to 2048 threads per CU would use 304 · 2048 · 8192 bytes = 4.75 GiB.

Reviewer: ```suggestion
to 2048 threads per CU could use 304 · 2048 · 8192 bytes = 4.75 GB for the call stack.
```

Reviewer: Also, this calculation seems to imply the stack size is dependent on the individual threads, even though previously we said it was based on wavefronts?

Author: It's a theoretical maximum. I added it the equivalent maximum in wavefronts.

Reviewer 2: I don't know what you tried to calculate here, but the numbers definitely don't add up. The max number of wavefronts on a MI300X is 304 CUs * 32 wavefronts / CU = 9728.

However that is not really relevant here: Randy's actual question is valid - the section about the call stack comparing NVIDIA and AMD GPUs would definitely benefit from a rewrite or it could even be removed, as currently it's just causing confusion.

Each and every thread needs a call stack, as they have individual variables (if the threads all had the same local variables, why even have multiple threads?), so the call stack needs to be stored per thread, for any and every GPU architecture.

The difference between NVIDIA and AMD GPUs is how the call stack is managed: pre independent-thread-scheduling NVIDIA GPUs' and AMD GPUs' wavefronts' progress in lock-step, meaning all threads in a wavefront take the same path, and don't progress independently. The drawback is obviously that non-participating threads have to be masked out when a function call is executed, and they have to wait for the other threads to return from the function, while the benefit lies in the hardware implementation - you only need one pointer to the call stack (the call stack still exists per thread!).
Independent thread scheduling enables the threads to progress independently, so that they don't have to wait for other threads in a warp to return from a function, but now the thread scheduler has to take care, that every thread knows where it has to return to, and that it can load its call stack frame independently. This requires an individual per-thread pointer to the call stack. This increases the management data needed for the call stack.

I'm not sure if it's worth it to add this explanation here, as it probably requires a more elaborate explanation and examples, for a small detail of the implementation.

Reviewer 3: Removed the calculation and the previous sentence too.

Reviewer 2: You also removed the correct calculation (the one that showed how much memory is taken up by the call stack). I think that one provided valuable information, as it showed why GPUs don't have a huge call stack and don't enable deep nested function calls by default - not because it is "complex" or for performance reasons (at least not more than any other function call on a GPU), but because it takes a lot of memory just for keeping track of the function calls!

Also the paragraph comparing the call stack of AMD GPUs and NVIDIA GPUs with independent thread scheduling is still there. As I pointed out earlier, that one either needs a complete overhaul, as it is quite confusing right now, or it should just be removed.

Handling recursion and deep function calls
-------------------------------------------------------------------------------

Similar to CPU programming, recursive functions and deeply nested function

> calls are supported. However, developers must ensure that these functions do

Reviewer: How do developers ensure these functions don't exceed the available stack memory? Error handling as shown below? Increase the stack size as shown above? Is there anything else? It seems to me this is simply suggesting that you use error handling to catch errors due to stack overflow?

Author: Thanks for the recommendation.
Both stack size increasing and error handling. But it also should include code optimization and the use of profilers.
I updated this section.

Reviewer 2: I don't know what you tried to calculate here, but the numbers definitely don't add up. The max number of wavefronts on a MI300X is 304 CUs * 32 wavefronts / CU = 9728.

However that is not really relevant here: Randy's actual question is valid - the section about the call stack comparing NVIDIA and AMD GPUs would definitely benefit from a rewrite or it could even be removed, as currently it's just causing confusion.

Each and every thread needs a call stack, as they have individual variables (if the threads all had the same local variables, why even have multiple threads?), so the call stack needs to be stored per thread, for any and every GPU architecture.

The difference between NVIDIA and AMD GPUs is how the call stack is **managed**: pre independent-thread-scheduling NVIDIA GPUs' and AMD GPUs' wavefronts' progress in lock-step, meaning all threads in a wavefront take the same path, and don't progress independently. The drawback is obviously that non-participating threads have to be masked out when a function call is executed, and they have to wait for the other threads to return from the function, while the benefit lies in the hardware implementation - you only need one **pointer** to the call stack (the call stack still exists per thread!).
Independent thread scheduling enables the threads to progress independently, so that they don't have to wait for other threads in a warp to return from a function, but now the thread scheduler has to take care, that every thread knows where it has to return to, and that it can load its call stack frame independently. This requires an individual per-thread **pointer** to the call stack. This increases the management data needed for the call stack.

I'm not sure if it's worth it to add this explanation here, as it probably requires a more elaborate explanation and examples, for a small detail of the implementation.

*******************************************************************************
Call stack
*******************************************************************************

The call stack is a data structure for managing function calls, by saving the
state of the current function. Each time a function is called, a new call frame
is added to the top of the stack, containing information such as local
variables, return addresses and function parameters. When the function
execution completes, the frame is removed from the stack and loaded back into
the corresponding registers. This concept allows the program to return to the
calling function and continue execution from where it left off.

The call stack for each thread must track its function calls, local variables,
and return addresses. However, in GPU programming, the memory required to store
the call stack increases due to the parallelism inherent to the GPUs. NVIDIA
and AMD GPUs use different approaches. NVIDIA GPUs have the independent thread
scheduling feature where each thread has its own call stack and effective
program counter. On AMD GPUs threads are grouped; each warp has its own call

> stack and program counter, but not each thread. Warps are described and 

Reviewer 2: ```suggestion
stack and program counter. Warps are described and 
```

.. meta::
    :description: This page lists graph safe ROCm libraries.

>     :keywords: AMD, ROCm, HIP, hipGRAPH

Reviewer 2: No, it doesn't :P

Author: Thanks for the catch! Fixed.

********************************************************************************
Call stack
********************************************************************************

The call stack and the program counter work together to manage the flow of
control in a program, especially in the context of function calls, loops, and

> branching.

Reviewer 2: I'm not sure if the program counter needs to be explicitly mentioned here. Program addresses are obviously important to the call stack, but the program counter is a somewhat different concept.

To get more specific: The call stack is not needed for loops and branching, so this might give the wrong impression of what it's needed for.

Author: Yes, you are right. I removed this paragraph.

Reviewer 2: The paragraph is still here

Author: True. Fixed.

The call stack is a concept in computer science, representing the stack data

> structure that stores information about the active subroutines or functions.

Reviewer 2: Wording:
```suggestion
The call stack stores information about the active subroutines or functions.
```
But also: I'd argue it doesn't store infromation about the **active** functions, but the ones that called the currently executed functions. As a concept it is used for managing function calls, by saving the state of the current function

Author: Updated. Can you check it, please?

> Each time a function is called, a new frame is added to the top of the stack,

Reviewer 2: Also applies to other occurences
```suggestion
Each time a function is called, a new call frame is added to the top of the stack,
```

Author: Added

containing information such as local variables, return addresses and function
parameters. When the function execution completes, the frame is removed from the

> stack. This concept allows the program to return to the calling function and

Reviewer 2: I think it's worth highlighting, that the frame is not just "removed", but loaded back into the corresponding registers. Otherwise why even keep it?

Author: Yes, it was not well phrased. Updated. Thanks.

> continue executing from where it left off.

Reviewer 2: ```suggestion
continue execution from where it left off.
```

Author: Updated

The program counter keeps track of the current or next instruction to

> be executed, while the call stack maintains the history of function calls,

Reviewer 2: This is a repetition of the previous paragraph

Author: You are right. This part of the sentence was removed and the paragraph was rephased.

Reviewer 2: This is still a repetition of the previous paragraph. The only information added when compared to the previous paragraph is "The program counter keeps track of the current or next instruction to be executed" which is also somehow out of place, as the program counter has not been mentioned before and is not really needed to describe the concept of the call stack.

Author: I removed this paragraph.

return addresses, and local variables. Together, they allow a program to execute
complex control flow structures, handle function calls and return to the

> appropriate location in the code after a function completes.

Reviewer 2: Control flow does not need a call stack. And while it's true, that both of this is needed, the program counter is somewhat implied - without the program counter no program can be run. The call stack is what is needed to enable function calls with returns to the appropriate location.

Author: I think the complex control flow structures were related to the next part, but it's really confusing, so I removed.

The call stack in GPU programming has extra complexity due to parallel execution
also NVIDIA and AMD GPUs are using different approaches. NVIDIA GPUs have the
independent thread scheduling feature since Volta GPU, where every thread has
its own call stack and effective program counter. At AMD GPUs every

> warp/wavefront has its own call stack and program counter.

Reviewer 2: I wouldn't say it increases complexity, the concept is still the same, but the memory required to store the call stack increases due to the parallelism inherent to the GPUs.

Also, when thinking about it: Does the difference in how NVIDIA and AMD GPUs handle the call stack matter that much? In the end, every thread has to have its own state stored on the stack (as long as it's not something that is constant across the wavefront). So the only difference is, that each thread for an NVIDIA GPU has its own stack pointer, but other than that the amount of memory needed to store the state (disregarding stack management data) is the same.

Author: Updated.
I would highlight the difference, perhaps some users can have a better intuition about it.

Reviewer 2: You are right, the difference is worth highlighting, however currently it reads as if this would somehow affect the size of the call stack for the different implementations ("one is per thread, the other per wavefront"), which is not entirely true. Both implementations still need to store the variables etc per thread, it's just that on NVIDIA every thread has an "individual" call stack, whereas on AMD they are grouped per wavefront.

Author: I rephased a bit. Can you check it, please?

The call stack for each thread must track its function calls, local variables,

> and return addresses.

Reviewer 2: This is again somewhat redundant. I'd explicitly integrate it into the previous paragraph, stating that that's the difference to a CPU call stack (however, that's not even really true - on the CPU also every thread has to have a call stack. There are just way fewer threads).

Author: Yes, you are right. I updated. Can you check it, please? How is it now?

If a thread or warp exceeds its stack size, a stack overflow occurs, causing
kernel failure. This can be detected and handled using debuggers or with the

> :doc:`error handling functions <../how-to/hip_runtime_api/error_handling>`.

Reviewer 2: I couldn't find an error type that relates to any stack errors. There is one for CUDA (cudaErrorHardwareStackError), but it seems to be unused in HIP.
Also: Debuggers might be able to detect them, but not handle them. Tools like valgrind can be used for that, however I'm still not sure how far along the GPU support for valgrind is

Author: HIP has no supported equivalent for `cudaErrorHardwareStackError`. The reference is removed.
I agree about the debuggers, I updated this part of the sentence too.

Call stack management with HIP
================================================================================

> Developers can adjust the call stack size for every thread, allowing fine-tuning

Reviewer 2: This can be slightly misunderstood - the stack size is the same for every thread, but it is set device-wide, so not individually for each thread.

Maybe just omit the "for every thread".

Author: Thanks. You are right. Updated.

based on specific kernel requirements. This helps prevent stack overflow errors
by ensuring sufficient stack memory is allocated.

.. code-block:: cpp
        
    #include <hip/hip_runtime.h>
    #include <iostream>

    #define HIP_CHECK(expression)                \
    {                                            \
        const hipError_t status = expression;    \
        if(status != hipSuccess){                \
                std::cerr << "HIP error "        \
                    << status << ": "            \
                    << hipGetErrorString(status) \
                    << " at " << __FILE__ << ":" \
                    << __LINE__ << std::endl;    \
        }                                        \
    }

    int main()
    {
        size_t stackSize;
        HIP_CHECK(hipDeviceGetLimit(&stackSize, hipLimitStackSize));
        std::cout << "Default stack size: " << stackSize << " bytes" << std::endl;

        // Set a new stack size

>         size_t newStackSize = 1024 * 8; // 1MB

Reviewer 2: ```suggestion
        size_t newStackSize = 1024 * 8; // 8 KiB
```
Might also be worth highlighting, that depending on what GPU this is, at full occupancy this can take up quite some memory. I.e. a MI300X with 304 CUs and possibly up to 2048 threads residing on each CU, this would take up 304 * 2048 * 8192 Bytes = 4.75 GiB

Author: Thanks. Updated. 
I like your example, I added it to the end of the code.

        HIP_CHECK(hipDeviceSetLimit(hipLimitStackSize, newStackSize));

>         std::cout << "New stack size set to: " << newStackSize << " bytes" << std::endl;

Reviewer 2: This line can be ommited. It just prints the variable that has been set "manually", not by hipDeviceSetLimit. The next print gives the actual information.

Author: I would keep it as a comparison between the wished and updated stack size. Perhaps we can change the order of the two lines. What do you think about it?

Reviewer 2: I don't see what information it provides - the desired stack size is not some random variable that comes from an unknown source, it literally is what the developer specifies beforehand.

The only way I could see that this provides additional information is when the function fails, and in that case the HIP_CHECK macro would provide that information

Author: Yes, the macro should manage it. Removed.

        HIP_CHECK(hipDeviceGetLimit(&stackSize, hipLimitStackSize));
        std::cout << "Updated stack size: " << stackSize << " bytes" << std::endl;

        return 0;
    }

Handling recursion and deep function calls
--------------------------------------------------------------------------------

Similar to CPU programming, recursive functions and deeply nested function calls
are supported. However, developers must ensure that these functions do not

> exceed the available stack memory, considering the limited resources on GPUs.

Reviewer 2: Maybe not "necessarily limited resources on GPUs", but the huge amount of memory needed for the call stack due to the GPUs inherent parallelism.

Author: Thanks. Updated.

.. code-block:: cpp

    #include <hip/hip_runtime.h>
    #include <iostream>
    
    #define HIP_CHECK(expression)                \
    {                                            \
        const hipError_t status = expression;    \
        if(status != hipSuccess){                \
                std::cerr << "HIP error "        \
                    << status << ": "            \
                    << hipGetErrorString(status) \
                    << " at " << __FILE__ << ":" \
                    << __LINE__ << std::endl;    \
        }                                        \
    }
    
    __device__ unsigned long long factorial(unsigned long long n)
    {
        if (n == 0 || n == 1) {
            return 1;
        }
        return n * factorial(n - 1);
    }
    
    __global__ void kernel(unsigned long long n)
    {
        unsigned long long result = factorial(n);
        const size_t x = threadIdx.x + blockDim.x * blockIdx.x;
        

>         if( x == 0)

Reviewer 2: 😠
```suggestion
        if(x == 0)
```

Author: Fixed.

          printf("%llu! = %llu \n", n, result);
    }
    
    int main()
    {
        kernel<<<1, 1>>>(10);

>         HIP_CHECK(hipDeviceSynchronize());

Reviewer 2: For some reason this didn't return an error for me, even when setting it to an unreasonable high number (the program crashed instead).

Also, while this is a nice example, it's not entirely suitable. 21! is already big enough to not fit in a 64 bit integer any more. Using doubles somewhat makes this better, but that way the result still reaches "infinity" before the stack overflow occurs.

Author: I'm thinking about it. Thanks!

Reviewer 2: As a suggestion: https://en.wikipedia.org/wiki/1_%2B_2_%2B_3_%2B_4_%2B_%E2%8B%AF

Author: Good idea!
I added the Fibonacci sequence. What do you think, is it okay, or should I add triangular numbers instead?

>         // With -O0 optimization option hit the stack limit
>         // kernel<<<1, 256>>>(2048);

Reviewer 2: Also hits the stack limit with optimization, and is independent of block size, so it should use the same config as the first call
```suggestion
        // This hits the stack limit due to excessive recursion
        // kernel<<<1, 1>>>(2048);
```

Author: Thanks! Updated.

Reviewer 2: I tested again on a different GPU, seems like this is somewhat unpredictable depending on the architecture, however with -O0 as you proposed earlier it "successfully" fails even on different architectures. Sorry, that was my bad, should have checked before with different GPUs.

Author: Reversed. Kept the original one.

        }                                        \
    }

    int main()
    {
        size_t stackSize;
        HIP_CHECK(hipDeviceGetLimit(&stackSize, hipLimitStackSize));
        std::cout << "Default stack size: " << stackSize << " bytes" << std::endl;

        // Set a new stack size
        size_t newStackSize = 1024 * 8; // 8 KiB
        HIP_CHECK(hipDeviceSetLimit(hipLimitStackSize, newStackSize));
        std::cout << "New stack size set to: " << newStackSize << " bytes" << std::endl;

        HIP_CHECK(hipDeviceGetLimit(&stackSize, hipLimitStackSize));
        std::cout << "Updated stack size: " << stackSize << " bytes" << std::endl;

        return 0;
    }

Depending on the GPU model, at full occupancy, it can consume a significant
amount of memory. For instance, an MI300X with 304 computing units (CU) and up
to 2048 threads per CU would use 304 · 2048 · 8192 bytes = 4.75 GiB.

Handling recursion and deep function calls
--------------------------------------------------------------------------------

Similar to CPU programming, recursive functions and deeply nested function calls
are supported. However, developers must ensure that these functions do not

> exceed the available stack memory, considering a huge amount of memory needed

Reviewer 2: ```suggestion
exceed the available stack memory, considering the huge amount of memory needed
```

Author: Fixed.
